src/transforms/aws_ec2_metadata.rs

```diff
@@ -0,0 +1,596 @@
+
+    static ref DYNAMIC_DOCUMENT: PathAndQuery = PathAndQuery::from_static("/latest/dynamic/instance-identity/document");
+
+    static ref DEFAULT_FIELD_WHITELIST: Vec<Atom> = vec![
+        AMI_ID_KEY.clone(),
+        AVAILABILITY_ZONE_KEY.clone(),
+        INSTANCE_ID_KEY.clone(),
+        LOCAL_HOSTNAME_KEY.clone(),
+        LOCAL_IPV4_KEY.clone(),
+        PUBLIC_HOSTNAME_KEY.clone(),
+        PUBLIC_IPV4_KEY.clone(),
+        REGION_KEY.clone(),
+        SUBNET_ID_KEY.clone(),
+        VPC_ID_KEY.clone(),
+        ROLE_NAME_KEY.clone(),
+    ];
+
+    static ref API_TOKEN: PathAndQuery = PathAndQuery::from_static("/latest/api/token");
+    static ref TOKEN_HEADER: Bytes = Bytes::from("X-aws-ec2-metadata-token");
+    static ref HOST: Uri = Uri::from_static("http://169.254.169.254");
+}
+
+#[derive(Default, Debug, Serialize, Deserialize)]
+pub struct Ec2Metadata {
+    host: Option<String>,
+    namespace: Option<String>,
+    refresh_interval_secs: Option<u64>,
+    fields: Option<Vec<String>>,
+}
+
+pub struct Ec2MetadataTransform {
+    state: ReadHandle,
+}
+
+#[derive(Debug, Clone)]
+struct Keys {
+    ami_id_key: Atom,
+    availability_zone_key: Atom,
+    instance_id_key: Atom,
+    local_hostname_key: Atom,
+    local_ipv4_key: Atom,
+    public_hostname_key: Atom,
+    public_ipv4_key: Atom,
+    region_key: Atom,
+    subnet_id_key: Atom,
+    vpc_id_key: Atom,
+    role_name_key: Atom,
+}
+
+inventory::submit! {
+    TransformDescription::new_without_default::<Ec2Metadata>("aws_ec2_metadata")
+}
+
+#[typetag::serde(name = "aws_ec2_metadata")]
+impl TransformConfig for Ec2Metadata {
+    fn build(&self, exec: TaskExecutor) -> crate::Result<Box<dyn Transform>> {
+        let (read, write) = evmap::new();
+
+        let keys = Keys::new(&self.namespace);
+        let host = self
+            .host
+            .clone()
+            .map(|s| Uri::from_shared(s.into()).unwrap())
+            .unwrap_or(HOST.clone());
+
+        let refresh_interval = self
+            .refresh_interval_secs
+            .map(Duration::from_secs)
+            .unwrap_or(Duration::from_secs(10));
+        let fields = self
+            .fields
+            .clone()
+            .map(|v| v.into_iter().map(Atom::from).collect())
+            .unwrap_or_else(|| DEFAULT_FIELD_WHITELIST.clone());
+
+        exec.spawn_std(
+            async move {
+                let mut client = MetadataClient::new(host, keys, write, refresh_interval, fields);
+
+                // Keep retrying the client if there are errors.
+                loop {
+                    if let Err(error) = client.fetch_metadata().await {
+                        error!(message = "Unable to refresh metadata.", %error);
+
+                        Delay::new(Instant::now() + Duration::from_secs(5))
+                            .compat()
+                            .await
+                            .unwrap();
+                    } else {
+                        break;
```

Comment:
Is this currently unreachable? It'd be nice if we could clearly see that this task can't quietly finish or die and leave the transform with stale data, but this `break` is implying that it can. I don't think we need to go full OTP supervisor for this, but we should try to structure things to make it clear what can happen.